File: shared/main/java/net/ccbluex/liquidbounce/file/configs/AccountsConfig.java
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.file.configs;

import java.io.BufferedWriter;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import com.google.gson.*;

import net.ccbluex.liquidbounce.file.FileConfig;
import net.ccbluex.liquidbounce.file.FileManager;
import net.ccbluex.liquidbounce.utils.ClientUtils;
import net.ccbluex.liquidbounce.utils.login.MinecraftAccount;
import net.ccbluex.liquidbounce.utils.login.MinecraftAccount.AltServiceType;
import net.ccbluex.liquidbounce.utils.misc.MiscUtils;

public class AccountsConfig extends FileConfig
{
	private final List<MinecraftAccount> accounts = new ArrayList<>();

	/**
	 * Constructor of config
	 *
	 * @param file
	 *             of config
	 */
	public AccountsConfig(final File file)
	{
		super(file);
	}

	/**
	 * Load config from file
	 *
	 * @throws IOException
	 */
	@SuppressWarnings("unchecked")
	@Override
	protected void loadConfig() throws IOException
	{
		clearAccounts();
		try
		{
			final JsonElement jsonElement = new JsonParser().parse(MiscUtils.createBufferedFileReader(getFile()));

			if (jsonElement instanceof JsonNull)
				return;

			final JsonArray jsonArray = jsonElement.getAsJsonArray();

			for (final JsonElement accountElement : jsonArray)
			{
				final JsonObject accountObject = accountElement.getAsJsonObject();
				final JsonElement type = accountObject.get("type");
				final JsonElement name = accountObject.get("name");
				final JsonElement password = accountObject.get("password");
				final JsonElement inGameName = accountObject.get("inGameName");
				final JsonElement bannedServers = accountObject.get("bannedServers");

				final MinecraftAccount account;
				if ((inGameName == null || inGameName.isJsonNull()) && !(password == null || password.isJsonNull()))
					account = new MinecraftAccount(AltServiceType.getById(type.getAsString()), name.getAsString(), password.getAsString());
				else if (type == null || type.isJsonNull())
					account = new MinecraftAccount(AltServiceType.MOJANG, name.getAsString());
				else
					account = (inGameName == null || inGameName.isJsonNull()) && (password == null || password.isJsonNull()) ? new MinecraftAccount(AltServiceType.getById(type.getAsString()), name.getAsString()) : new MinecraftAccount(AltServiceType.getById(type.getAsString()), name.getAsString(), password.getAsString(), inGameName.getAsString());

				if (!bannedServers.isJsonNull())
				{
					final List<String> servers = new ArrayList<>();
					for (final JsonElement element : bannedServers.getAsJsonArray())
						servers.add(element.getAsString());
					account.setBannedServers(servers);
				}

				addAccount(account);
			}

		}
		catch (final JsonSyntaxException | IllegalStateException ex)
		{
			// When the JSON Parse fail, the client try to load and update the old config
			ClientUtils.getLogger().info("[FileManager] Try to load old Accounts config...");
			final Iterable<String> accountList = new Gson().fromJson(MiscUtils.createBufferedFileReader(getFile()), List.class);

			if (accountList == null)
				return;

			accounts.clear();

			for (final String account : accountList)
			{
				final String[] info = account.split("/", 2);
				if (info.length > 0)
				{
					final String[] information = info[0].split(":");

					final MinecraftAccount acc;

					switch (information.length)
					{
						case 1:
							acc = new MinecraftAccount(AltServiceType.MOJANG, information[0]);
							break;
						case 2:
							acc = new MinecraftAccount(AltServiceType.getById(information[0]), information[1]);
							break;
						case 3:
							acc = new MinecraftAccount(AltServiceType.getById(information[0]), information[1], information[2]);
							break;
						default:
							acc = new MinecraftAccount(AltServiceType.getById(information[0]), information[1], information[2], information[3]);
					}

					if (info.length > 1)
						acc.setBannedServers(deserializeOldBannedServers(info[1]));

					accounts.add(acc);
				}
			}
			ClientUtils.getLogger().info("[FileManager] Loaded old Accounts config...");

			// Save the accounts into a new valid JSON file
			saveConfig();
			ClientUtils.getLogger().info("[FileManager] Saved Accounts to new config...");
		}

	}

	/**
	 * Save config to file
	 *
	 * @throws IOException
	 */
	@Override
	protected void saveConfig() throws IOException
	{
		final JsonArray jsonArray = new JsonArray();

		for (final MinecraftAccount minecraftAccount : accounts)
		{
			final JsonObject accountObject = new JsonObject();
			accountObject.addProperty("type", minecraftAccount.getServiceType().getId());
			accountObject.addProperty("name", minecraftAccount.getName());
			accountObject.addProperty("password", minecraftAccount.getPassword());
			accountObject.addProperty("inGameName", minecraftAccount.getAccountName());
			if (!minecraftAccount.getBannedServers().isEmpty())
			{
				final JsonArray arr = new JsonArray();
				for (final String server : minecraftAccount.getBannedServers())
					arr.add(new JsonPrimitive(server));
				accountObject.add("bannedServers", arr);
			}
			jsonArray.add(accountObject);
		}

		final BufferedWriter writer = MiscUtils.createBufferedFileWriter(getFile());
		writer.write(FileManager.PRETTY_GSON.toJson(jsonArray) + System.lineSeparator());
		writer.close();
	}

	public static List<String> deserializeOldBannedServers(final String str)
	{
		final String[] split = str.split(";");
		return new ArrayList<>(Arrays.asList(split));
	}

	/**
	 * Add cracked account to config
	 *
	 * @param account
	 *                The account
	 */
	public void addAccount(final MinecraftAccount account)
	{
		if (isAccountExists(account.getName()))
			return;

		accounts.add(account);
	}

	/**
	 * Remove account from config
	 *
	 * @param selectedSlot
	 *                     of the account
	 */
	public void removeAccount(final int selectedSlot)
	{
		accounts.remove(selectedSlot);
	}

	/**
	 * Removed an account from the config
	 *
	 * @param account
	 *                the account
	 */
	public void removeAccount(final MinecraftAccount account)
	{
		accounts.remove(account);
	}

	/**
	 * Check if the account is already added
	 *
	 * @param  name
	 *              of account
	 * @return      if the account exists
	 */
	public boolean isAccountExists(final String name)
	{
		return accounts.stream().anyMatch(minecraftAccount -> minecraftAccount.getName().equals(name));
	}

	/**
	 * Clear all minecraft accounts from alt array
	 */
	public void clearAccounts()
	{
		accounts.clear();
	}

	/**
	 * Get Alts
	 *
	 * @return list of Alt Accounts
	 */
	public List<MinecraftAccount> getAccounts()
	{
		return accounts;
	}
}

File: shared/main/java/net/ccbluex/liquidbounce/file/configs/AccountsConfig.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.file.configs

import com.google.gson.*
import net.ccbluex.liquidbounce.file.FileConfig
import net.ccbluex.liquidbounce.file.FileManager
import net.ccbluex.liquidbounce.utils.ClientUtils.logger
import net.ccbluex.liquidbounce.utils.login.MinecraftAccount
import net.ccbluex.liquidbounce.utils.login.MinecraftAccount.AltServiceType
import net.ccbluex.liquidbounce.utils.misc.MiscUtils
import java.io.File
import java.io.IOException
import java.util.*
import java.util.regex.Pattern

/**
 * Constructor of config
 *
 * @param file
 * of config
 */
class AccountsConfig(file: File) : FileConfig(file)
{
	val accounts: MutableList<MinecraftAccount> = ArrayList()

	// TODO: private val lock = ReentrantReadWriteLock()

	private val slashPattern: Pattern = Pattern.compile("/", Pattern.LITERAL)

	/**
	 * Load config from file
	 *
	 * @throws IOException
	 */
	@Throws(IOException::class)
	override fun loadConfig()
	{
		clearAccounts()
		try
		{
			val jsonElement = JsonParser().parse(MiscUtils.createBufferedFileReader(file))
			if (jsonElement is JsonNull) return
			val jsonArray = jsonElement.asJsonArray
			for (accountElement in jsonArray)
			{
				val accountObject = accountElement.asJsonObject
				val type: JsonElement? = accountObject["type"]
				val name: JsonElement = accountObject["name"]
				val password: JsonElement? = accountObject["password"]
				val inGameName: JsonElement? = accountObject["inGameName"]
				val bannedServers: JsonElement? = accountObject["bannedServers"]

				val account: MinecraftAccount = if (inGameName?.isJsonNull != false && password?.isJsonNull == false)
				{
					val serviceType = type?.let { AltServiceType.getById(it.asString) } ?: AltServiceType.MOJANG
					MinecraftAccount(serviceType, name.asString, password.asString)
				}
				else if (type?.isJsonNull != false) MinecraftAccount(AltServiceType.MOJANG, name.asString)
				else
				{
					val serviceType = AltServiceType.getById(type.asString) ?: AltServiceType.MOJANG
					if ((inGameName?.isJsonNull != false) && (password?.isJsonNull != false)) MinecraftAccount(serviceType, name.asString) else MinecraftAccount(serviceType, name.asString, password?.asString, inGameName?.asString)
				}

				if (bannedServers?.isJsonNull == false) account.bannedServers = bannedServers.asJsonArray.mapTo(ArrayList()) { it.asString }

				// lock.readLock().lock()
				addAccount(account)
				// lock.readLock().unlock()
			}
		}
		catch (ex: JsonSyntaxException)
		{
			loadWithOldFormat(ex)
		}
		catch (ex: IllegalStateException)
		{
			loadWithOldFormat(ex)
		}
	}

	private fun loadWithOldFormat(ex: Exception)
	{
		// When the JSON Parse fail, the client try to load and update the old config
		logger.info("[FileManager] Failed to load Accounts config with new format. Try to load old Accounts config...", ex)

		val accountList = Gson().fromJson<List<*>>(MiscUtils.createBufferedFileReader(file), MutableList::class.java) ?: return

		accounts.clear()

		accountList.map { it.toString().split(slashPattern, 2) }.filter(List<String>::isNotEmpty).forEach { info ->
			val information = info[0].split(":")
			val serviceType = AltServiceType.getById(information[0]) ?: AltServiceType.MOJANG

			val acc: MinecraftAccount = when (information.size)
			{
				1 -> MinecraftAccount(AltServiceType.MOJANG, information[0])
				2 -> MinecraftAccount(serviceType, information[1])
				3 -> MinecraftAccount(serviceType, information[1], information[2])
				else -> MinecraftAccount(serviceType, information[1], information[2], information[3])
			}

			if (info.size > 1) acc.bannedServers = deserializeOldBannedServers(info[1])

			accounts.add(acc)
		}

		logger.info("[FileManager] Loaded old Accounts config...")

		// Save the accounts into a new valid JSON file
		saveConfig()
		logger.info("[FileManager] Saved Accounts to new config...")
	}

	/**
	 * Save config to file
	 *
	 * @throws IOException
	 */
	@Throws(IOException::class)
	override fun saveConfig()
	{
		val jsonArray = JsonArray()

		// lock.writeLock().lock()
		for (minecraftAccount in accounts)
		{
			val accountObject = JsonObject()
			accountObject.addProperty("type", minecraftAccount.serviceType.id)
			accountObject.addProperty("name", minecraftAccount.name)
			accountObject.addProperty("password", minecraftAccount.password)
			accountObject.addProperty("inGameName", minecraftAccount.accountName)

			val bannedServers = minecraftAccount.bannedServers
			if (bannedServers.isNotEmpty())
			{
				val arr = JsonArray()
				bannedServers.forEach(arr::add)
				accountObject.add("bannedServers", arr)
			}

			jsonArray.add(accountObject)
		}
		// lock.writeLock().lock()

		val writer = MiscUtils.createBufferedFileWriter(file)
		writer.write(FileManager.PRETTY_GSON.toJson(jsonArray) + System.lineSeparator())
		writer.close()
	}

	/**
	 * Add cracked account to config
	 *
	 * @param account
	 * The account
	 */
	fun addAccount(account: MinecraftAccount)
	{
		if (isAccountExists(account.name)) return
		accounts.add(account)
	}

	// /**
	//  * Remove account from config
	//  *
	//  * @param selectedSlot
	//  * of the account
	//  */
	// fun removeAccount(selectedSlot: Int)
	// {
	// 	accounts.removeAt(selectedSlot)
	// }

	/**
	 * Removed an account from the config
	 *
	 * @param account
	 * the account
	 */
	fun removeAccount(account: MinecraftAccount?)
	{
		accounts.remove(account)
	}

	/**
	 * Check if the account is already added
	 *
	 * @param  name
	 * of account
	 * @return      if the account exists
	 */
	fun isAccountExists(name: String): Boolean = accounts.any { name.equals(it.name, ignoreCase = true) }

	/**
	 * Clear all minecraft accounts from alt array
	 */
	private fun clearAccounts()
	{
		accounts.clear()
	}

	companion object
	{
		fun deserializeOldBannedServers(str: String): MutableList<String> = str.split(";").toMutableList()
	}
}
